File: src/main/java/mygame/editor/parser/LocalParser.java
package mygame.editor.parser;

import com.badlogic.gdx.math.Vector2;
import mygame.editor.model.Node;
import mygame.editor.model.Physics;
import mygame.editor.model.Point;
import mygame.editor.model.Sprite;
import mygame.editor.model.box2d.B2Fixture;
import mygame.editor.model.box2d.B2FixtureType;
import mygame.editor.model.box2d.B2Type;
import mygame.editor.views.*;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class LocalParser {

    private static final String NODES = "nodes";
    private static final String ID = "id";
    private static final String X = "x";
    private static final String Y = "y";
    private static final String NAME = "name";
    private static final String WIDTH = "width";
    private static final String HEIGHT = "height";
    private static final String ANGLE = "angle";
    private static final String IMAGE = "image";
    private static final String TYPE = "type";
    private static final String SHAPE = "shape";
    private static final String NODE = "node";
    private static final String SPRITE = "sprite";
    private static final String PHYSICS = "physics";
    private static final String FIXTURES = "fixtures";
    private static final String RESTITUTION = "restitution";
    private static final String DENSITY = "density";
    private static final String FRICTION = "friction";
    private static final String POINTS = "points";

    private static final String CENTER = "center";
    private static final String RADIUS = "radius";

    private static final String RECT = "rect";
    private static final String CIRCLE = "circle";
    private static final String CHAIN = "chain";
    private static final String POLYGON = "polygon";
    private static final String DYNAMIC = "dynamic";
    private static final String KINEMATIC = "kinematic";
    private static final String STATIC = "static";


    public static List<Node> createNodesFromString(String json) {

        List<Node> layers = new ArrayList<>();


        JSONObject jsonObject = new JSONObject(json);
        JSONArray nodes = jsonObject.getJSONArray(NODES);

        for (int i = 0; i < nodes.length(); i++) {
            JSONObject jsonNode = nodes.getJSONObject(i);
            Node node = parseNode(jsonNode);
            layers.add(node);
        }

        return layers;
    }

    public static String createJsonFromNodes(List<Node> nodes) {
        JSONObject jsonObject = new JSONObject();
        JSONArray jsonArray = new JSONArray();
        jsonObject.put(NODES, jsonArray);
        for (Node ccNode : nodes) {
            JSONObject jsonNode = createJsonNode(ccNode);
            jsonArray.put(jsonNode);
        }
        return jsonObject.toString();
    }

    private static Node parseNode(JSONObject jsonObject) {
        String type = jsonObject.getString(TYPE);
        if (SPRITE.equals(type)) {
            return createSprite(jsonObject);
        } else {
            return createNode(jsonObject);
        }

    }

    private static Node createNode(JSONObject jsonObject) {
        Node node = new Node();
        fillNode(node, jsonObject);
        return node;
    }

    private static Sprite createSprite(JSONObject jsonObject) {
        String image = jsonObject.getString(IMAGE);
        Sprite sprite = new Sprite(image);
        fillNode(sprite, jsonObject);
        return sprite;
    }

private static void fillNode(Node node, JSONObject jsonObject) {
        double x = jsonObject.getDouble(X);
        double y = jsonObject.getDouble(Y);
        double width = jsonObject.getDouble(WIDTH);
        double height = jsonObject.getDouble(HEIGHT);
        String name = jsonObject.getString(NAME);
        double angle = jsonObject.getDouble(ANGLE);

        node.getPosition().set(x,y);
        node.getWidth().set(width);
        node.getHeight().set(height);
        node.getName().set(name);
        node.getAngle().set(angle);
        if (jsonObject.has(PHYSICS)) {

            JSONObject physicsObject = jsonObject.getJSONObject(PHYSICS);

            String physicsType = physicsObject.getString(TYPE);
            double friction = physicsObject.getDouble(FRICTION);
            double restitution = physicsObject.getDouble(RESTITUTION);
            double density = physicsObject.getDouble(DENSITY);

            Physics physics = new Physics();
            physics.setType(physicsType);
            physics.setFriction(friction);
            physics.setRestitution(restitution);
            physics.setDensity(density);

            String shape = physicsObject.getString(SHAPE);

            switch (shape){
                case RECT:
                    double rectWidth = physicsObject.getDouble(WIDTH);
                    double rectHeight = physicsObject.getDouble(HEIGHT);
                    physics.setWidth(rectWidth);
                    physics.setHeight(rectHeight);
                    break;
                case CIRCLE :
                    JSONObject centerPointJson = physicsObject.getJSONObject(CENTER);
                    double centerPointX = centerPointJson.getDouble(X);
                    double centerPointY = centerPointJson.getDouble(Y);
                    double radius  = physicsObject.getDouble(RADIUS);
                    physics.setCenter(centerPointX,centerPointY);
                    physics.setRadius(radius);
                    break;
                case CHAIN :
                    JSONArray pointsJson = physicsObject.getJSONArray(POINTS);
                    for (int i = 0; i < pointsJson.length(); i++) {
                        JSONObject jsonPoint = pointsJson.getJSONObject(i);
                        double pointX = jsonPoint.getDouble(X);
                        double pointY = jsonPoint.getDouble(Y);
                        physics.getPoints().add(new Point(pointX, pointY));
                    }
                    break;

            }
            node.setPhysics(physics);

        }

        if (jsonObject.has(NODES)) {
            JSONArray nodes = jsonObject.getJSONArray(NODES);
            for (int index = 0;index < nodes.length(); index++){
                JSONObject jsonNode = nodes.getJSONObject(index);
                Node parsedNode = parseNode(jsonNode);
                node.addChild(parsedNode);
            }
        }
    }



    private static JSONObject createJsonNode(Node node) {

        JSONObject json;
        if (node instanceof Sprite){
            json = createJsonSprite((Sprite) node);
        } else{
            json = createJsonDefaultNode(node);

        }

        if (node.getPhysics() != null) {
            JSONObject editBody = createPhysicsJson(node.getPhysics());
            json.put(PHYSICS, editBody);
        }


        JSONArray children = new JSONArray();
        for (Node child :node.getChildren()) {
            JSONObject jsonChild = createJsonNode(child);
            children.put(jsonChild);
        }
        json.put(NODES, children);
        return json;


    }

    private static JSONObject createPhysicsJson(Physics physics) {

        JSONObject jsonObject = new JSONObject();

        jsonObject.put(RESTITUTION, physics.getRestitution().get());
        jsonObject.put(DENSITY, physics.getDensity().get());
        jsonObject.put(FRICTION, physics.getFriction().get());
        jsonObject.put(TYPE,physics.getType());
        switch (physics.getShape().getValue()){
            case RECT:
                jsonObject.put(SHAPE,RECT);
                jsonObject.put(WIDTH, physics.getWidth().doubleValue());
                jsonObject.put(HEIGHT,physics.getHeight().doubleValue());
                break;
            case CIRCLE:
                jsonObject.put(SHAPE,CIRCLE);
                jsonObject.put(RADIUS,physics.getRadius().doubleValue());
                JSONObject center = new JSONObject();
                center.put(X,physics.getCenter().getX().doubleValue());
                center.put(Y,physics.getCenter().getY().doubleValue());
                jsonObject.put(CENTER,center);

                break;
            case CHAIN:
                jsonObject.put(SHAPE,CHAIN);
                JSONArray points = new JSONArray();
                for (Point point : physics.getPoints()) {
                    JSONObject p = new JSONObject();
                    p.put(X, point.getX().getValue());
                    p.put(Y, point.getY().getValue());
                    points.put(p);
                }
                jsonObject.put(POINTS, points);
                break;
        }

        return jsonObject;
    }

    private static JSONObject createJsonSprite(Sprite node) {
        JSONObject jsonObject = createJsonDefaultNode(node);
        jsonObject.put(TYPE, SPRITE);
        jsonObject.put(IMAGE, node.getImage().get());
        return jsonObject;
    }

    private static JSONObject createJsonDefaultNode(Node node) {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put(TYPE, NODE);
        jsonObject.put(X, node.getPosition().getX().doubleValue());
        jsonObject.put(Y, node.getPosition().getY().doubleValue());
        jsonObject.put(NAME, node.getName().getValue());
        jsonObject.put(WIDTH, node.getWidth().doubleValue());
        jsonObject.put(HEIGHT, node.getHeight().doubleValue());
        jsonObject.put(ANGLE, node.getAngle().doubleValue());
        return jsonObject;
    }
}

File: src/main/java/mygame/editor/parser/LocalParser.kt
package mygame.editor.parser

import com.badlogic.gdx.math.Vector2
import mygame.editor.model.box2d.B2Body
import mygame.editor.model.box2d.B2Fixture
import mygame.editor.model.box2d.B2FixtureType
import mygame.editor.model.box2d.B2Type
import mygame.editor.views.*
import org.json.JSONArray
import org.json.JSONObject


private const val NODES = "nodes"
private const val ID = "id"
private const val X = "x"
private const val Y = "y"
private const val NAME = "name"
private const val WIDTH = "width"
private const val HEIGHT = "height"
private const val ANGLE = "angle"
private const val IMAGE = "image"
private const val TYPE = "type"
private const val NODE = "node"
private const val SPRITE = "sprite"
private const val PHYSICS = "physics"
private const val FIXTURES = "fixtures"
private const val RESTITUTION = "restitution"
private const val DENSITY = "density"
private const val FRICTION = "friction"
private const val POINTS = "points"

private const val RECT = "rect"
private const val CIRCLE = "circle"
private const val CHAIN = "chain"
private const val EDGE = "edge"
private const val POLYGON = "polygon"

private const val DYNAMIC = "dynamic"
private const val KINEMATIC = "kinematic"
private const val STATIC = "static"


fun createNodesFromString(json: String): List<CcNode> {

    val layers = arrayListOf<CcNode>()


    val jsonObject = JSONObject(json)
    val nodes = jsonObject.getJSONArray(NODES)

    for (i in 0 until nodes.length()) {
        val jsonNode = nodes.getJSONObject(i)
        val node = parseNode(jsonNode)
        layers.add(node)
    }

    return layers
}

fun createJsonFromNodes(nodes: List<CcNode>): String {
    val jsonObject = JSONObject()
    val jsonArray = JSONArray()
    jsonObject.put(NODES, jsonArray)
    for (ccNode in nodes) {
        val jsonNode = createJsonNode(ccNode)
        jsonArray.put(jsonNode)
    }
    return jsonObject.toString()
}

private fun parseNode(jsonObject: JSONObject): CcNode {
    val type = jsonObject.getString(TYPE)
    return when (type) {
        SPRITE -> createSprite(jsonObject)
        else -> createNode(jsonObject)
    }
}

private fun createNode(jsonObject: JSONObject): CcNode {
    val node = CcNode()
    fillNode(node, jsonObject)
    return node
}

private fun createSprite(jsonObject: JSONObject): CcSprite {
    val image = jsonObject.getString(IMAGE)
    val sprite = CcSprite(image)
    fillNode(sprite, jsonObject)
    return sprite
}

private fun fillNode(node: CcNode, jsonObject: JSONObject) {
    val x = jsonObject.getDouble(X)
    val y = jsonObject.getDouble(Y)
    val width = jsonObject.getDouble(WIDTH)
    val height = jsonObject.getDouble(HEIGHT)
    val name = jsonObject.getString(NAME)
    val angle = jsonObject.getDouble(ANGLE)

    node.x.set(x)
    node.y.set(y)
    node.width.set(width)
    node.height.set(height)
    node.name.set(name)
    node.angle.set(angle)
    if (jsonObject.has(PHYSICS)) {

        val physicsObject = jsonObject.getJSONObject(PHYSICS)
        val physicsType = physicsObject.getString(TYPE)
        val bodyType = parseBodyType(physicsType)
        val body = B2Body(bodyType)

        val fixturesArray = physicsObject.getJSONArray(FIXTURES)
        for (index in 0 until fixturesArray.length()) {
            val fixtureJson = fixturesArray.getJSONObject(index)
            val fixtureType = fixtureJson.getString(TYPE)


            val fixtureTypeEnum = parseFixtureType(fixtureType)


            val friction = fixtureJson.getFloat(FRICTION)
            val restitution = fixtureJson.getFloat(RESTITUTION)
            val density = fixtureJson.getFloat(DENSITY)

            val pointsJsonArray = fixtureJson.getJSONArray(POINTS)

            val list = arrayListOf<Vector2>()
            for (pointIndex in 0 until pointsJsonArray.length()) {
                val jsonPoint = pointsJsonArray.getJSONObject(pointIndex)
                val pointX = jsonPoint.getFloat(X)
                val pointY = jsonPoint.getFloat(Y)
                list.add(Vector2(pointX, pointY))

            }
            val fixture = B2Fixture(fixtureTypeEnum, list)
            fixture.density = density
            fixture.restitution = restitution
            fixture.friction = friction

            body.addFixture(fixture)
        }
        node.editBody = CcEditBodyNode(body)

    }

    if (jsonObject.has(NODES)) {
        val nodes = jsonObject.getJSONArray(NODES)
        for (index in 0 until nodes.length()) {
            val jsonNode = nodes.getJSONObject(index)
            val parsedNode = parseNode(jsonNode)
            node.addChild(parsedNode)
        }
    }
}

private fun parseBodyType(type:String):B2Type{
    return when(type){
        STATIC ->B2Type.STATIC
        DYNAMIC ->B2Type.DYNAMIC
        else->B2Type.KINEMATIC
    }
}

private fun parseFixtureType(type: String): B2FixtureType {
    return when (type) {
        RECT -> B2FixtureType.RECT
        CIRCLE -> B2FixtureType.CIRCLE
        CHAIN -> B2FixtureType.CHAIN
        EDGE -> B2FixtureType.CHAIN
        else -> B2FixtureType.POLYGON
    }
}

private fun enumToBodyType(enum :B2Type):String{
    return when(enum){
        B2Type.STATIC -> STATIC
        B2Type.DYNAMIC -> DYNAMIC
        else -> KINEMATIC
    }
}

private fun enumToFixtureType(enum: B2FixtureType):String{
    return when(enum){
        B2FixtureType.CHAIN-> CHAIN
        B2FixtureType.CIRCLE -> CIRCLE
        B2FixtureType.RECT -> RECT
        B2FixtureType.EDGE -> EDGE
        else -> POLYGON
    }
}
private fun createJsonNode(node: CcNode): JSONObject? {

    val json: JSONObject
    if (node is CcSprite) {
        json = createJsonSprite(node)
    } else {
        json = createJsonDefaultNode(node)

    }

    if(node.editBody != null){
        val editBody = createPhysicsJson(node.editBody);
        json.put(PHYSICS,editBody)
    }


    val children = JSONArray()
    for (child in node.children) {
        val jsonChild = createJsonNode(child)
        children.put(jsonChild)
    }
    json.put(NODES, children)
    return json


}

fun createPhysicsJson(editBody: CcEditBodyNode): JSONObject {
    val jsonObject = JSONObject()
    val b2EditBody = editBody.b2EditBody
    jsonObject.put(TYPE, enumToBodyType(b2EditBody.type))

    jsonObject.put(ID,b2EditBody.id)
    val jsonFixtures = JSONArray()
    jsonObject.put(FIXTURES,jsonFixtures)
    for (fixtureNode in editBody.fixtureNodes) {
        val jsonFixture = JSONObject()
        val fixture = fixtureNode.fixture
        jsonFixture.put(RESTITUTION,fixture.restitution)
        jsonFixture.put(DENSITY,fixture.density)
        jsonFixture.put(FRICTION,fixture.friction)
        val fixtureType = enumToFixtureType(fixture.type)
        jsonFixture.put(TYPE,fixtureType)

        if(fixtureNode is CcPolygon){
            jsonFixture.put(TYPE, POLYGON)
        }else if(fixtureNode is CcCircle){
            jsonFixture.put(TYPE, CIRCLE)
        }else if(fixture is CcChain){
            jsonFixture.put(TYPE, CHAIN)
        }

    }



    return jsonObject
}

private fun createJsonSprite(node: CcSprite): JSONObject {
    val jsonObject = JSONObject()
    jsonObject.put(TYPE, SPRITE)
    jsonObject.put(X, node.x.doubleValue())
    jsonObject.put(Y, node.y.doubleValue())
    jsonObject.put(NAME, node.name.value)
    jsonObject.put(WIDTH, node.width.doubleValue())
    jsonObject.put(HEIGHT, node.height.doubleValue())
    jsonObject.put(ANGLE, node.angle.doubleValue())
    jsonObject.put(IMAGE, node.imagePath)
    return jsonObject
}

private fun createJsonDefaultNode(node: CcNode): JSONObject {
    val jsonObject = JSONObject()
    jsonObject.put(TYPE, NODE)
    jsonObject.put(X, node.x.doubleValue())
    jsonObject.put(Y, node.y.doubleValue())
    jsonObject.put(NAME, node.name.value)
    jsonObject.put(WIDTH, node.width.doubleValue())
    jsonObject.put(HEIGHT, node.height.doubleValue())
    jsonObject.put(ANGLE, node.angle.doubleValue())
    return jsonObject
}

